opencxl/cxl/component/cxl_mem_dcoh.py

```diff
@@ -0,0 +1,380 @@
+    ):
+        self._downstream_fifo: Optional[FifoPair]
+        self._upstream_fifo: FifoPair
+
+        super().__init__(upstream_fifo, downstream_fifo, label)
+        self._cache_to_coh_agent_fifo = cache_to_coh_agent_fifo
+        self._coh_agent_to_cache_fifo = coh_agent_to_cache_fifo
+        self._memory_device_component: Optional[CxlMemoryDeviceComponent] = None
+
+        # snoop filter defined as set structure
+        # max sf size will be the same as each cache size
+        self._sf_host = set()
+        self._sf_device = set()
+        self._addr = 0
+
+    def set_memory_device_component(self, memory_device_component: CxlMemoryDeviceComponent):
+        self._memory_device_component = memory_device_component
+
+    def _snoop_filter_update(self, addr, sf_update_list) -> None:
+        for sf_type in sf_update_list:
+            if sf_type == SF_UPDATE_TYPE.SF_HOST_IN:
+                self._sf_host.add(addr)
+            elif sf_type == SF_UPDATE_TYPE.SF_HOST_OUT:
+                self._sf_host.remove(addr)
+            elif sf_type == SF_UPDATE_TYPE.SF_DEVICE_IN:
+                self._sf_device.add(addr)
+            elif sf_type == SF_UPDATE_TYPE.SF_DEVICE_OUT:
+                self._sf_device.remove(addr)
+
+    def _sf_host_is_hit(self, addr) -> bool:
+        return addr in self._sf_host
+
+    def _sf_device_is_hit(self, addr) -> bool:
+        return addr in self._sf_device
+
+    # .mem response may have two packets (nds and drs)
+    # this method always makes two reponse packets but caller may ignore one if possible
+    def _create_mem_rsp_packet(
+        self,
+        ndr_opcode: CXL_MEM_S2MNDR_OPCODE,
+        data: Optional[int] = 0,
+        drs_opcode: Optional[CXL_MEM_S2MDRS_OPCODE] = CXL_MEM_S2MDRS_OPCODE.MEM_DATA,
+        meta_field: Optional[CXL_MEM_META_FIELD] = CXL_MEM_META_FIELD.NO_OP,
+        meta_value: Optional[CXL_MEM_META_VALUE] = CXL_MEM_META_VALUE.ANY,
+    ) -> Tuple[CxlMemCmpPacket, CxlMemMemDataPacket]:
+        return (
+            CxlMemCmpPacket.create(ndr_opcode, meta_field, meta_value),
+            CxlMemMemDataPacket.create(data, drs_opcode, meta_field, meta_value),
+        )
+
+    # .mem m2s req (MemRd, MemRdData, and MemInv) handler
+    async def _process_cxl_m2s_req_packet(self, m2sreq_packet: CxlMemM2SReqPacket):
+        if self._downstream_fifo is not None:
+            logger.debug(self._create_message("Forwarding CXL.mem M2S Req packet"))
+            await self._downstream_fifo.host_to_target.put(m2sreq_packet)
+            return
+
+        if self._memory_device_component is None:
+            raise Exception("CxlMemoryDeviceComponent isn't set yet")
+
+        addr = m2sreq_packet.get_address()
+
+        if m2sreq_packet.m2sreq_header.meta_field == CXL_MEM_META_FIELD.NO_OP:
+            data = await self._memory_device_component.read_mem(addr)
+
+            dummy, packet = self._create_mem_rsp_packet(CXL_MEM_S2MNDR_OPCODE.CMP, data)
```

Comment:
Python convention for dummy is `_`

`_, packet = self._create_mem_rsp_packet(CXL_MEM_S2MNDR_OPCODE.CMP, data)`